File: rust/src/providers/parsers.rs
use anyhow::{bail, Context, Result};
use chrono::{DateTime, Utc};
use serde_json::Value;

use crate::domain::UsageWindow;

#[derive(Debug, Clone)]
pub struct ParsedWindows {
    pub observed_at: DateTime<Utc>,
    pub windows: Vec<UsageWindow>,
}

pub fn parse_timestamp(val: &Value, milliseconds: bool) -> Result<DateTime<Utc>> {
    if let Some(n) = val.as_i64() {
        let secs = if milliseconds { n / 1000 } else { n };
        let nsecs = if milliseconds { ((n % 1000) * 1_000_000) as u32 } else { 0 };
        return DateTime::from_timestamp(secs, nsecs)
            .context("invalid unix timestamp");
    }
    if let Some(f) = val.as_f64() {
        let secs = if milliseconds { (f / 1000.0) as i64 } else { f as i64 };
        let nsecs = if milliseconds { ((f % 1000.0) * 1_000_000.0) as u32 } else { 0 };
        return DateTime::from_timestamp(secs, nsecs)
            .context("invalid unix float timestamp");
    }
    if let Some(s) = val.as_str() {
        if let Ok(dt) = DateTime::parse_from_rfc3339(s) {
            return Ok(dt.with_timezone(&Utc));
        }
        let fixed = s.replace('Z', "+00:00");
        if let Ok(dt) = DateTime::parse_from_rfc3339(&fixed) {
            return Ok(dt.with_timezone(&Utc));
        }
    }
    bail!("unsupported timestamp format: {:?}", val)
}

pub fn parse_codex_rollout(lines: &[&str]) -> Result<ParsedWindows> {
    let mut newest: Option<(DateTime<Utc>, Vec<UsageWindow>)> = None;

    for line in lines {
        let Ok(event): Result<Value, _> = serde_json::from_str(line) else {
            continue;
        };
        let Some(ts_val) = event.get("timestamp") else {
            continue;
        };
        let Ok(ts) = parse_timestamp(ts_val, false) else {
            continue;
        };

        let Some(raw_limits) = event.pointer("/payload/rate_limits").and_then(|v| v.as_object()) else {
            continue;
        };

        let mut windows = Vec::new();
        for (key, val) in raw_limits {
            if (key == "primary" || key == "secondary") && val.is_object() {
                let used = val.get("used_percent").or_else(|| val.get("usedPercent")).and_then(|v| v.as_f64()).unwrap_or(0.0);
                let duration = val.get("window_minutes").or_else(|| val.get("windowDurationMins")).and_then(|v| v.as_i64());
                let reset = val.get("resets_at").or_else(|| val.get("resetsAt")).and_then(|v| parse_timestamp(v, false).ok());

                windows.push(UsageWindow {
                    id: key.clone(),
                    used_percent: used,
                    window_minutes: duration,
                    resets_at: reset,
                    raw_limits: None,
                });
            }
        }

        if !windows.is_empty() && newest.as_ref().is_none_or(|(prev_ts, _)| ts > *prev_ts) {
            newest = Some((ts, windows));
        }
    }

    if let Some((ts, windows)) = newest {
        Ok(ParsedWindows {
            observed_at: ts,
            windows,
        })
    } else {
        bail!("no valid Codex rate limit event found")
    }
}

pub fn parse_antigravity_quota_summary(document: &Value) -> Result<ParsedWindows> {
    let response = document.get("response").unwrap_or(document);
    let groups = response.get("groups").and_then(|v| v.as_array()).context("no groups in quota summary")?;

    let mut windows = Vec::new();
    for group in groups {
        let group_name = group.get("displayName").and_then(|v| v.as_str()).unwrap_or("").to_lowercase();
        let group_id = if group_name.contains("gemini") { "gemini" } else { "third_party" };

        if let Some(buckets) = group.get("buckets").and_then(|v| v.as_array()) {
            for bucket in buckets {
                let remaining = match bucket.get("remainingFraction").and_then(|v| v.as_f64()) {
                    Some(r) if (0.0..=1.0).contains(&r) => r,
                    _ => continue,
                };
                let bucket_id = bucket.get("bucketId").and_then(|v| v.as_str()).unwrap_or("").to_lowercase();
                let window_kind = bucket.get("window").and_then(|v| v.as_str()).unwrap_or("").to_lowercase();

                let (suffix, duration) = if bucket_id.contains("weekly") || window_kind == "weekly" {
                    ("weekly", Some(10080))
                } else if bucket_id.contains("5h") || window_kind == "5h" || window_kind == "five_hour" || window_kind == "five-hour" {
                    ("five_hour", Some(300))
                } else {
                    (bucket_id.as_str(), None)
                };

                let resets_at = bucket.get("resetTime").and_then(|v| parse_timestamp(v, false).ok());

                windows.push(UsageWindow {
                    id: format!("{}_{}", group_id, suffix.replace('-', "_")),
                    used_percent: ((1.0 - remaining) * 10000.0).round() / 100.0,
                    window_minutes: duration,
                    resets_at,
                    raw_limits: None,
                });
            }
        }
    }

    if windows.is_empty() {
        bail!("no valid quota buckets found");
    }

    Ok(ParsedWindows {
        observed_at: Utc::now(),
        windows,
    })
}

File: archive/python/src/time_to_sleep/providers/parsers.py
import json
import re
from collections.abc import Iterable, Mapping
from datetime import UTC, datetime, timedelta
from typing import Any

from time_to_sleep.domain import UsageWindow
from time_to_sleep.providers.base import ParsedWindows


def _timestamp(value: Any, *, milliseconds: bool = False) -> datetime:
    if not isinstance(value, (int, float, str)):
        raise ValueError("timestamp is invalid")
    if isinstance(value, (int, float)):
        return datetime.fromtimestamp(float(value) / (1000 if milliseconds else 1), tz=UTC)
    text = value.replace("Z", "+00:00")
    parsed = datetime.fromisoformat(text)
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=UTC)
    return parsed.astimezone(UTC)


def _percent(value: Any) -> float:
    if not isinstance(value, (int, float)) or not 0 <= float(value) <= 100:
        raise ValueError("usage percentage is invalid")
    return float(value)


def _window(window_id: str, raw: Mapping[str, Any]) -> UsageWindow:
    used = raw.get("used_percent", raw.get("usedPercent"))
    duration = raw.get("window_minutes", raw.get("windowDurationMins"))
    reset = raw.get("resets_at", raw.get("resetsAt"))
    return UsageWindow(
        id=window_id,
        used_percent=_percent(used),
        window_minutes=int(duration) if duration is not None else None,
        resets_at=_timestamp(reset) if reset is not None else None,
    )


def parse_codex_rollout(lines: Iterable[str]) -> ParsedWindows:
    newest: tuple[datetime, Mapping[str, Any]] | None = None
    for line in lines:
        try:
            event = json.loads(line)
            timestamp = _timestamp(event.get("timestamp"))
            raw_limits = event.get("payload", {}).get("rate_limits")
            if not isinstance(raw_limits, Mapping):
                continue
            windows = {
                key: value
                for key, value in raw_limits.items()
                if key in {"primary", "secondary"} and isinstance(value, Mapping)
            }
            if windows and (newest is None or timestamp > newest[0]):
                newest = timestamp, windows
        except (AttributeError, KeyError, TypeError, ValueError, json.JSONDecodeError):
            continue

    if newest is None:
        raise ValueError("no valid Codex rate limit event found")
    timestamp, raw_windows = newest
    return ParsedWindows(
        observed_at=timestamp,
        windows=[_window(window_id, raw) for window_id, raw in raw_windows.items()],
    )


def parse_claude_plan_history(document: Mapping[str, Any]) -> ParsedWindows:
    samples = document.get("samples")
    if not isinstance(samples, list):
        raise ValueError("Claude plan history has no valid samples")

    valid_samples: list[tuple[datetime, Mapping[str, Any]]] = []
    for sample in samples:
        if not isinstance(sample, Mapping):
            continue
        try:
            timestamp = _timestamp(sample["t"], milliseconds=True)
            usage = sample["u"]
            if not isinstance(usage, Mapping):
                continue
            valid_samples.append((timestamp, usage))
        except (KeyError, TypeError, ValueError):
            continue

    if not valid_samples:
        raise ValueError("Claude plan history has no valid samples")
    observed_at, usage = max(valid_samples, key=lambda item: item[0])
    windows = []
    for window_id, source_key, duration in (
        ("five_hour", "fh", 300),
        ("seven_day", "sd", 10080),
    ):
        if source_key in usage:
            windows.append(
                UsageWindow(
                    id=window_id,
                    used_percent=_percent(usage[source_key]),
                    window_minutes=duration,
                )
            )
    if not windows:
        raise ValueError("Claude plan history has no valid usage windows")
    return ParsedWindows(observed_at=observed_at, windows=windows)


_ANTIGRAVITY_EVENT = re.compile(
    r"(?P<timestamp>\d{4}-\d{2}-\d{2}(?:T| )\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z)?)"
    r"[\s\S]{0,240}?Individual quota reached[\s\S]{0,160}?"
    r"Resets in (?P<hours>\d+)h(?P<minutes>\d+)m(?P<seconds>\d+)s",
    re.IGNORECASE,
)


def parse_antigravity_log(content: str, now: datetime | None = None) -> ParsedWindows:
    current = (now or datetime.now(UTC)).astimezone(UTC)
    matches = list(_ANTIGRAVITY_EVENT.finditer(content))
    if not matches:
        raise ValueError("no Antigravity quota event found")

    latest = matches[-1]
    observed_at = _timestamp(latest.group("timestamp"))
    duration = timedelta(
        hours=int(latest.group("hours")),
        minutes=int(latest.group("minutes")),
        seconds=int(latest.group("seconds")),
    )
    resets_at = observed_at + duration
    if resets_at <= current:
        raise ValueError("Antigravity quota event is stale")
    return ParsedWindows(
        observed_at=observed_at,
        windows=[
            UsageWindow(
                id="quota-event",
                used_percent=100,
                window_minutes=max(1, round(duration.total_seconds() / 60)),
                resets_at=resets_at,
            )
        ],
    )


def parse_antigravity_quota_summary(
    document: Mapping[str, Any], now: datetime | None = None
) -> ParsedWindows:
    response = document.get("response", document)
    if not isinstance(response, Mapping):
        raise ValueError("Antigravity quota summary response is not an object")
    groups = response.get("groups")
    if not isinstance(groups, list):
        raise ValueError("Antigravity quota summary has no groups")

    windows: list[UsageWindow] = []
    for group in groups:
        if not isinstance(group, Mapping):
            continue
        group_name = str(group.get("displayName", "")).lower()
        group_id = "gemini" if "gemini" in group_name else "third_party"
        buckets = group.get("buckets")
        if not isinstance(buckets, list):
            continue
        for bucket in buckets:
            if not isinstance(bucket, Mapping):
                continue
            remaining = bucket.get("remainingFraction")
            if not isinstance(remaining, (int, float)) or not 0 <= float(remaining) <= 1:
                continue
            bucket_id = str(bucket.get("bucketId", "")).lower()
            window_kind = str(bucket.get("window", "")).lower()
            if "weekly" in bucket_id or window_kind == "weekly":
                suffix = "weekly"
                duration = 10080
            elif "5h" in bucket_id or window_kind in {"5h", "five_hour", "five-hour"}:
                suffix = "five_hour"
                duration = 300
            else:
                suffix = bucket_id.replace("-", "_") or "unknown"
                duration = None
            reset = bucket.get("resetTime")
            try:
                resets_at = _timestamp(reset) if reset is not None else None
            except ValueError:
                resets_at = None
            windows.append(
                UsageWindow(
                    id=f"{group_id}_{suffix}",
                    used_percent=round((1 - float(remaining)) * 100, 4),
                    window_minutes=duration,
                    resets_at=resets_at,
                )
            )

    if not windows:
        raise ValueError("Antigravity quota summary has no valid buckets")
    observed_at = (now or datetime.now(UTC)).astimezone(UTC)
    return ParsedWindows(observed_at=observed_at, windows=windows)
